keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,633 @@
+
+- `PeriodMilliseconds`       // How often (in milliseconds) to offset PeriodSeconds when performing the probe.
+- `InitialDelayMilliseconds` // Length of time (in milliseconds) to offset IntialDelaySeconds before health checking is activated.
+- `TimeoutMilliseconds`      // Length of time (in milliseconds) to offset TimeoutMilliseconds before health checking times out.
+
+The seconds and milliseconds fields (as related) will be summed to get the resulting time duration. For example,
+
+```yaml
+...
+#yaml
+periodSeconds: 1
+periodMilliseconds: 500
+...
+```
+
+would be considered a period value of 1.5s or 1500ms. These values are used as `time.Duration` within the prober package ([ex1](https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/pkg/kubelet/prober/prober.go#L161) [ex2](https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/pkg/kubelet/prober/worker.go#L132) ), so there's no real difference between 1.5s and 1500ms.
+
+A few additional examples:
+
+`periodSeconds: 2` and `periodMilliseconds: -500` would be 1.5s / 1500ms.
+
+`periodSeconds: 1` and `periodMilliseconds: -900` would be 0.1s / 100ms.
+
+More generally, the effective period value would be = `periodSeconds` + `periodMilliseconds`.
+
+There are a few corner cases around default (effective period) values:
+
+`periodSeconds: 0` and `periodMilliseconds: 500` would be 10.5s / 10500ms (as 0 => 10 for `periodSeconds` via [defaults.go](https://github.com/kubernetes/kubernetes/blob/3b13e9445a3bf86c94781c898f224e6690399178/pkg/apis/core/v1/defaults.go#L213-L215))
+
+`timeoutSeconds: 0` and `timeoutMilliseconds: 500` would be 1.5s / 1500ms (as 0 => 1 for `timeoutSeconds` via [defaults.go](https://github.com/kubernetes/kubernetes/blob/3b13e9445a3bf86c94781c898f224e6690399178/pkg/apis/core/v1/defaults.go#L210-L212))
+
+Each optional millisecond field will be restricted to [-999,999], and the effective sum will never be allowed
+to be less than 0 (if the effective sum is less than 0, it will fail at the validation stage and block delployments).
+
+
+### User Stories (Optional)
+
+#### Knative
+
+See the discussion from the Sept 14, 2021 SIG-Node meeting
+[recording](https://youtu.be/LMh7c9e7H-Q?list=PL69nYSiGNLP1wJPj5DYWXjiArF-MJ5fNG&t=1049)
+[meeting notes](https://docs.google.com/document/d/1Ne57gvidMEWXR70OxxnRkYquAoMpt56o75oZtg-OeBg/edit#heading=h.by9uk7onna00)
+
+### Notes/Constraints/Caveats (Optional)
+
+Changing defaults is a strict no-go.
+
+### Risks and Mitigations
+
+Accidentally setting a timeout too low could DOS kubelet if many are used.
+Will mitigate by preventing timeout values that are too small.
+Could be configurable,
+100ms is a first guess, but will be adjusted based on user feedback during alpha and performance testing. Will also implement a scaling of the periodMilliseconds field to further reduce risk of thrashing.
```

Comment:
I think exec vs. other types of probes will vary a lot in how much time they take to execute so need to look into that as well.

Reply:
nod.. perhaps based on type makes more sense..

Reply:
latest commit changes to ^